src/mahalanobis.jl

```diff
@@ -1,19 +1,26 @@
 # Mahalanobis distances
 
-struct Mahalanobis{T} <: Metric
-    qmat::Matrix{T}
+struct Mahalanobis{M<:AbstractMatrix} <: Metric
+    qmat::M
 end
 
-struct SqMahalanobis{T} <: SemiMetric
-    qmat::Matrix{T}
+struct SqMahalanobis{M<:AbstractMatrix} <: SemiMetric
+    qmat::M
 end
 
-result_type(::Mahalanobis{T}, ::Type, ::Type) where {T} = T
-result_type(::SqMahalanobis{T}, ::Type, ::Type) where {T} = T
+function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
+    z = oneunit(T1) - oneunit(T2)
+    return typeof(z * oneunit(eltype(d.qmat)) * z)
+end
+
+function result_type(d::SqMahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
+    z = oneunit(T1) - oneunit(T2)
+    return typeof(z * oneunit(eltype(d.qmat)) * z)
+end
```

Comment:
As mentioned above, an alternative would be to use `zero`
```suggestion
function result_type(d::SqMahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = zero(T1) - zero(T2)
    return typeof(z * zero(eltype(d.qmat)) * z)
end
```
since it might be more natural and more likely to be defined.